blog/2021-08-20-Exploring-Template-Haskell.md

```diff
@@ -0,0 +1,574 @@
+
+Since its conception nearly two decades ago [1], Template Haskell has enabled Haskellers to more easily manipulate their programs. Nowadays, TH is mostly used to:
+
+- Write different programs all at once (for example, several instances for a class) by means of a single algorithmic description, thus reducing 'boilerplate' code
+- Embed files or external code into Haskell modules
+
+In Template Haskell, in contrast to meta-programming in C++, the functions that execute at compile-time are written in the same language as the functions that execute at run time, that is, Haskell. This means that you can directly apply your Haskell skills and use standard libraries when writing TH programs.
+
+# How does Template Haskell work?
+
+To understand how Template Haskell works, we first need to have a clear idea of how programs are represented in Haskell. Haskell syntax can be roughly divided into four categories [2]:
+
+- **Expressions**
+    - These are the most common, and represent anything that GHC can evaluate: variables, literals, lambdas, function applications, if-then-else expressions, and so on.
+        - Generally, this corresponds to the right-hand-side of function definitions.
+- **Declarations**
+    - Declarations are simply statements that introduce something new. Some examples are:
+        - function declarations: `f x = x + 1` introduces a meaning to the variable `f`.
+        - class declarations: `class Eq a where ...`,
+        - instance declarations: `instance Eq Int ...` declares that `Int` is a member of the type class `Eq`,
+        - type declarations: `data Bool = False | True`, and
+        - the statements that come in a `where ...` clause.
+- **Patterns**
+    - Patterns are used to 'match' expressions, and they are widely used in Haskell functions. In function declarations, they are placed between the function name and the `=` sign.
+        - For example, in `f 4 = 5`, the literal `4` is a pattern against which the argument passed to `f` will be matched (successfully or not).
+        - Constructors can also be patterns: see `f (Just val) = val` or `f Nothing = 0`.
+- **Types**
+    - Types are the 'labels' or 'categories' of terms.
+        - Some examples are `Int`, `Bool -> Int`, `Monad m => a -> b -> m (c,d)`, etc.
+
+![](/images/blogposts/template-haskell/fibonacci_diagram.png)
+
+    It turns out that this simple classification allows us to express almost anything in Haskell and, by extension, Template Haskell! In fact, we can write the TH version of any ‘normal’ Haskell code by simply converting it into a TH abstract syntax tree, which is done by using the constructors of the Dec, Exp, Pat and Type data types defined in the [template-haskell](https://hackage.haskell.org/package/template-haskell) support library.
+
+    # Getting your hands dirty
```

Comment:
This might be issue in the Notion exporter, but it seems that everything below this point is indented, where it shouldn't. This breaks the formatting of the resulting page.
Can you un-indent everything below this line (e.g. using multi-line cursor in vscode Ctrl+Shift+arrow down and once all the lines are selected, press Shift+Tab to unindend)?